backend/routes/rankings.js

```diff
@@ -0,0 +1,138 @@
+const express = require("express");
+
+// allows routes to be sent out
+const router = express.Router();
+const db = require("../mongo.js");
+
+const { isAuthenticated } = require("../middlewares/auth");
+
+router.use(isAuthenticated);
+
+router.get("/:userid/:programId", function(req, res) {
+  db["Authentication"].rankings
+    .aggregate([
+      {
+        $match: { userId: req.params.userid, programId: req.params.programId }
+      },
+      {
+        $unwind: "$applications"
+      },
+      {
+        $lookup: {
+          from: "Reviews",
+          let: {
+            appId: "$applications.appId",
+            userId: "$userId",
+            programId: "$programId"
+          },
+          pipeline: [
+            {
+              $match: {
+                $expr: {
+                  $and: [
+                    { $eq: ["$applicationId", "$$appId"] },
+                    { $eq: ["$userId", "$$userId"] },
+                    { $eq: ["$programId", "$$programId"] }
+                  ]
+                }
+              }
+            },
+            {
+              $addFields: {
+                questionRatings: {
+                  $map: {
+                    input: "$sectionList",
+                    as: "item",
+                    in: {
+                      $cond: {
+                        if: { $gte: ["$$item.rating", 0] },
+                        then: "$$item.rating",
+                        else: null
+                      }
+                    }
+                  }
+                }
+              }
+            },
+            {
+              $project: {
+                rating: 1,
+                suggested: {
+                  $avg: "$questionRatings"
+                }
+              }
+            }
+          ],
+          as: "ratingInfo"
+        }
+      },
+      {
+        $lookup: {
+          from: "Applications",
+          let: { appId: "$applications.appId" },
+          pipeline: [
+            { $match: { $expr: { $eq: ["$_id", "$$appId"] } } },
+            {
+              $project: {
+                "Organization Name": 1,
+                "Organization Name (legal name)": 1
+              }
+            }
+          ],
+          as: "applicationInfo"
+        }
+      },
+      {
+        $replaceRoot: {
+          newRoot: {
+            $mergeObjects: [
+              "$$ROOT",
+              { $arrayElemAt: ["$ratingInfo", 0] },
+              { $arrayElemAt: ["$applicationInfo", 0] }
+            ]
+          }
+        }
+      },
+      {
+        $project: { applications: 0, applicationInfo: 0, ratingInfo: 0 }
+      }
+    ])
+    .then(function(found) {
+      res.json(found);
+    })
+    .catch(function(err) {
+      res.send(err);
+    });
+});
+
+//Admin stats
+router.get("/", function(req, res) {
+  db["Authentication"].rankings
+    .find()
+    .then(function(found) {
+      res.json(found);
+    })
+    .catch(function(err) {
+      res.send(err);
+    });
+});
```

Comment:
can we make this an endpoint based on the program ID? may be more performant instead of returning all ranking data to the client

Reply:
Yes. I actually still need to update this endpoint in the frontend, so good catch. 

btw, I am not about the life of return all and filter on the frontend. I am a changed man